### PR TITLE
chore(rpc): build rpc as `OnComponentsInitializedHook`

### PR DESCRIPTION
**Pull request by emhane**
Build RPC component as `OnComponentsInitializedHook`. Part of task to expose `EthApi` top level, for user to configure, like for different networks.

Exposes extension component builders as `Option<OnComponentsInitializedHook<N>>`s. The hook takes a mutable reference to the extension component container as param. This is how extension component builders stage their output for installation in the node, as well as how they read state from previously initialised components that they may need to build themselves. This has two main benefits:

- Since extension components are supposed to be optional, it's better that component inter-dependencies are not modelled with rigid type transitions, e.g. `Engine<Node>` is the only way to get to `Rpc<Engine<Node>>`, unlike for core components.
- This flexibility allows us to implement the extension component builders in several PRs. For this PR, engine is just staged from the scope of `DefaultNodeLauncher::launch_node`.

### CODE REVIEW

File: crates/node/builder/src/launch/common.rs [hunk 1]
```diff
@@ -843,6 +869,102 @@ where
...
+
+        NodeAdapterExt { core, rpc }
+    }
+}
+
+pub trait InstallPipeline: ExtComponentBuilder {}
+pub trait InstallBeaconEngine: ExtComponentBuilder + InstallPipeline {}
+pub trait InstallRpc: ExtComponentBuilder + InstallBeaconEngine {
+    // fn beacon_engine_handle(&self) -> &Self::Node::Engine::Handle;
+    fn beacon_engine_handle(&self) -> &BeaconConsensusEngineHandle<Self::Node::Engine>;
```
Comment (emhane) [inline crates/node/builder/src/launch/common.rs:937]: will just move this getter into `RpcBuilder` trait, and same for other dependencies getters, and delete these `Install` traits

File: crates/node/api/src/node.rs [hunk 2]
```diff
@@ -145,3 +163,19 @@ pub trait FullNodeComponents: FullNodeTypes + Clone + 'static {
...
+}
+
+pub trait Rpc<N: FullNodeComponents>: Send + Sync {
+    type ServerHandles;
+    type Registry;
+
+    fn handles(&self) -> &Self::ServerHandles;
+
+    fn registry(&self) -> &Self::Registry;
+}
```
Comment (mattsse) [inline crates/node/api/src/node.rs:181]: I think RPC is downstream of components and part of the `Node` / `Launch` process.

so I think we want:
* keep the rpc extension API on nodebuilder
* make rpc part of the Node helper trait, to embed it into `DefaultLauncher`

but somehow we'd need to a way to clarify the EthApi type for the nodehooks I assume?

maybe a solution that could work is is replacing the RpcHooks type with the RpcBuilder

Comment (mattsse) [reply crates/node/api/src/node.rs:181]: > we may also want to separate engine rpc and regular rpc as two components

definitely, this behaviour should be entirely node specific, hence I'm thinking this should be a separate thing and perhaps replace

https://github.com/paradigmxyz/reth/blob/38f2d00c8a25db4b46586ba00d8bbf27e6cff326/crates/node/builder/src/rpc.rs#L35-L35

and also be configured

https://github.com/paradigmxyz/reth/blob/38f2d00c8a25db4b46586ba00d8bbf27e6cff326/crates/node/builder/src/node.rs#L23-L23

**Comment by emhane** [comment]
closing, needs design

**Pull request closed without merging**